Implement offload to Celeritas from Geant4 using G4VTrackingManager hook

drbenmorgan opened the pull request:
This implements a minimal concrete class of the `G4VTrackingManager` interface for offloading Geant4 tracks to Celeritas, complementing the existing user tracking action and fast simulation strategies. Introduced in Geant4 11.0, a `G4VTrackingManager` instance can be added to one or more `G4ParticleDefinition` instances to customise how tracks of that particle the will be handled. Overviews of this capability, and an example are documented at these links:

- [Specialized Tracking Managers in Geant4](https://indico.cern.ch/event/1073798/contributions/4516033/attachments/2304502/3920446/specialized-tracking-manager.pdf)
- [G4HepEm and Specialized Stepping/Tracking in Geant4](https://indico.cern.ch/event/1052654/contributions/4524767/attachments/2309218/3929219/G4HepEm_SpecTracking_MNovak.pdf)
- [Geant4 runAndEvent RE07 Example](https://indico.cern.ch/event/1156193/contributions/5048470/attachments/2514883/4326508/20220927_Hahnfeld_RE07.pdf)
  - [Current implementation in Geant4 11.1](https://gitlab.cern.ch/geant4/geant4/-/tree/v11.1.3/examples/extended/runAndEvent/RE07?ref_type=tags)

Because implementation requires inheritance from a class only available from Geant4 v11.0, the build/use has been restricted to that version or newer for now.

The implementation here for the offload/flush is largely the same as in `FastSimulationOffload`, using `SharedParams` and `LocalTransporter`. 

The main detail is the necessity to implement the `Prepare/BuildPhysicsTable` member functions as per-particle tracking managers completely overrides physic building for the particle in question (see [here](https://gitlab.cern.ch/geant4/geant4/-/blob/v11.1.3/source/run/src/G4VUserPhysicsList.cc?ref_type=tags#L793-799), and [here](https://gitlab.cern.ch/geant4/geant4/-/blob/v11.1.3/source/run/src/G4VUserPhysicsList.cc?ref_type=tags#L644-658) in `G4VUserPhysicsList`). If we don't do this, no Geant4 physics data is available for that particle, so Celeritas can't construct its physics properly. `Prepare/BuildPhysicsTable` are therefore implemented to follow the default in `G4VUserPhysicsList`, as is also shown in [the runAndEvent/RE07 example](https://gitlab.cern.ch/geant4/geant4/-/blob/v11.1.3/examples/extended/runAndEvent/RE07/src/SpecializedTrackingManager.cc?ref_type=tags). _Aside: this rebuild of the physics data would also allow filtering in `HandOffOneTrack` to pass back to general tracking if certain preconditions are met, much like in fast simulation as well._

Like `FastSimulationOffload`, the current implementation does rely on the setup and use of `SimpleOffload`s Run and Event actions to properly setup/teardown params and transporter(s). An example of the use of `TrackingManagerOffload` is shown in a new program in `example/accel` which is identical to the others in problem setup, just using the new offload mechanism. This has also been documented.

Possible TODO: the example shows that the custom tracking manager is constructed and added to the particle definitions in the `ConstructProcess` member function. AFAICT this is only called on the master thread and thus with the `LocalTransporter` for that thread. However, local tests showed (via logging the address) that the correct thread-local instances are all used on the worker threads. The `G4VTrackingManager` instance held by `G4ParticleDefinition` _is_ part of the split data for that class, so seems to work correctly, but I'd still like to understand this better!

Addresses: #1035

drbenmorgan committed:
Implement concrete G4VTrackingManager for offload to Celeritas

Interface for offload of tracks to Celeritas via Geant4's G4VTrackingManager
hook introduced in v11.0, complementing existing fast simulation,
tracking action offload strategies.

Implementations of offload and flush largely follow those of
existing fast simulation offload. Build/PreparePhysicsTable are
overidden to ensure Geant4 tables for the particle(s) handled
by the tracking manager are rebuilt and thus available to
Celeritas to build its tables. This follows the Geant4 extended
runAndEvent/RE07 example.

Restrict build of class to Geant4 11.0 and higher as abstract
G4VTrackingManager interface is not available prior to that.

drbenmorgan committed:
Minimal example of using TrackingManagerOffload

Largely copy-paste of existing simple/fastsim offload examples, adapted
for new TrackingManagerOffload class. Primary changes are:

* Custom physics constructor inheriting from G4EmStandardPhysics
  * Overrides ConstructProcess, taking base class defaults except for
    constructing a TrackingManagerOffload instance and adding this
    to the electron, positron, and gamma particle definitions
* Replace EM physics in FTFP_BERT physics lists with this new physics
  constructor

drbenmorgan committed:
Add basic docs on TrackingManagerOffload

drbenmorgan committed:
Remove particle type validation

Prepare/BuildPhysicsTable have to be called before Celeritas'
SharedParams can be constructed, and so the list of offloadable
particles will be empty.

Remove validation check here as it cannot be run at this point.

sethrj commented on example/accel/trackingmanager-offload.cc:
```diff
@@ -0,0 +1,209 @@
+    using G4EmStandardPhysics::G4EmStandardPhysics;
+
+    void ConstructProcess() override
+    {
```
The physics constructor class instance is "shared", but is `ConstructProcess` done thread-locally? For the sake of clarity for the example I'd add:

```suggestion
    {
        CELER_LOG_LOCAL(status) << "Setting up tracking manager offload";
```

sethrj commented on example/accel/trackingmanager-offload.cc:
```diff
@@ -0,0 +1,209 @@
+            &shared_params, &local_transporter);
+        G4Electron::Definition()->SetTrackingManager(celer_tracking);
+        G4Positron::Definition()->SetTrackingManager(celer_tracking);
+        G4Gamma::Definition()->SetTrackingManager(celer_tracking);
```
Does anyone take ownership of this pointer or is it leaked?

drbenmorgan replied:
The deletion of the pointer is done in the destructor of `G4VUserPhysicsList` via the [`RemoveTrackingManager` member function](https://gitlab.cern.ch/geant4/geant4/-/blob/v11.1.3/source/run/src/G4VUserPhysicsList.cc?ref_type=tags#L326-356). That's analogous to what's done for process managers, albeit making the ownership non-definite.

sethrj commented on src/accel/TrackingManagerOffload.cc:
```diff
@@ -0,0 +1,141 @@
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a tracking manager with data needed to offload to Celeritas
```
For the "autobrief" doxygen to work, the first line here (and in all the rest of the docs) need a trailing full stop `.` .

sethrj commented on example/accel/CMakeLists.txt:
```diff
@@ -68,3 +78,12 @@ if(Geant4_VERSION VERSION_LESS 11.1)
+    ENVIRONMENT "${_test_env}"
+    LABELS "app"
+  )
+endif()
```
I wonder with the more complex threading going on, should we run these examples with serial/mt/task? (Task is the default in 11.)

drbenmorgan replied:
It certainly doesn't hurt - would just be a case of using the `G4FORCE_RUN_MANAGER_TYPE` env var with `Serial`, `MT`, `Tasking` to select the mode (so, three tests per app assuming an MT supporting install of Geant4). That's only going to be a smoke test though as there's otherwise no scoring or reference output check done.

drbenmorgan replied:
Sorry, realised my comment here was a bit vague - I'll go ahead and add these tests unless you say otherwise.

drbenmorgan replied:
Now done - it should work all the way back to Geant4 10.6, worst case 10.7.

sethrj replied:
Nice!! Great work, thanks @drbenmorgan !

sethrj commented on src/accel/TrackingManagerOffload.cc:
```diff
@@ -0,0 +1,141 @@
+    G4ProcessManager* pManager = part.GetProcessManager();
+    G4ProcessManager* pManagerShadow = part.GetMasterProcessManager();
+
+    G4ProcessVector* pVector = pManager->GetProcessList();
+    for (std::size_t j = 0; j < pVector->size(); ++j)
```
```suggestion
    CELER_ASSERT(pManager && pManager->GetProcessList());
    for (G4VProcess* process : *pManager->GetProcessList())
```
and below `(*pVector)[j]` -> `process`

drbenmorgan replied:
Despite its name, `G4ProcessVector` doesn't provide an interface compatible with, or trivially adaptable to, range-based for ☹️! We could write the loop as is done in `GeantImporter`:

```cpp
    CELER_ASSERT(pManager);
    G4ProcessVector* pVector = pManager->GetProcessList();
    CELER_ASSERT(pVector);
    for(auto j : range(pVector->size())
    {
        G4VProcess* p = *pVector[j];
        ....
```
?

drbenmorgan replied:
Implemented as above, which I think is the best that can be done without adding an iterator adaptor for `G4ProcessVector` (will add a Geant4 issue to review collection interfaces for "proper" behaviour!).

sethrj replied:
Ha, sounds good. Cheers!

sethrj reviewed (commented):
Looks great! So glad to have you helping us with this, as it looks like the thread abstractions are a real headache 😄

drbenmorgan committed:
Log call to ConstructProcess to clarify use on master/worker

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>

drbenmorgan committed:
Add trailing full stop to satisfy Doxygen autobrief

drbenmorgan committed:
Test offload examples in each Geant4 threading model

For each offload example, add a test that runs it with each of Geant4's
threading types:

- Serial
- MT (standard C++ threads)
- Tasking (PTL-based tasks+threadpool)

Only enable the latter two if the found Geant4 supports them.

drbenmorgan committed:
Clarify implementations of physics table functions

- Use contracts on pointers that require deref
- Follow GeantImporter implementation for loop over processes

A cleaner range-based for over processes isn't possible due to
G4ProcessVector not providing a directly compatible interface.

sethrj reviewed (approved):
Excellent!

sethrj committed:
Merge branch 'develop' into celer-tracking-manager
